src/table/boot.rs

```diff
@@ -752,9 +754,20 @@ impl BootServices {
         // `Status::BUFFER_TOO_SMALL` and `buffer` will be dropped.
         buffer.truncate(buffer_size);
 
+        // Convert the buffer from MaybeUninits to Handles
+        // the raw parts conversion is better than a transmute
+        let mut buffer = ManuallyDrop::new(buffer);
+        let handles = unsafe {
+            Vec::from_raw_parts(
+                buffer.as_mut_ptr() as *mut Handle,
+                buffer.len(),
+                buffer.capacity(),
+            )
+        };
```

Comment:
Do you have a link to something that verifies this pattern is safe? It looks right to me, but given the unsafe stuff going on with `MaybeUninit` and `Vec` I don't want to rely too much on my intuition for correctness. I would have thought that converting `Vec<MaybeUninit<T>>` to `Vec<T>` would be a common enough pattern to be googleable, but I failed to find it. Hopefully something does exist and I just missed it.

Reply:
According to this SO [anwser](https://stackoverflow.com/questions/48308759/how-do-i-convert-a-vect-to-a-vecu-without-copying-the-vector/55081958#55081958) and `transmute` docs, the `Vec::from_raw_parts` is the way to transmute vectors.

And then according to `MaybeUninit` docs if you are sure that the content is initialized you can transmute/reinterpret it (which is what `assume_init` does) - it has the same repr/alignment/etc as the target type. Can't quickly find the correct example/statement, but [here](https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#initializing-an-array-element-by-element) they transmute an array of maybe-uninits to an array of the target type, so this seems valid.

So I think this is correct

Reply:
In https://github.com/rust-lang/rfcs/pull/2756#issuecomment-866489007, and also in the SO answer you linked, they suggest using the (nightly-only) [`Vec::into_raw_parts`](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.into_raw_parts), which would then remove the need to use `ManuallyDrop`. A minor cleanup, but probably worth doing here since this crate is definitely nightly-only for now.

Reply:
Sure, did that, also added a shorter explanation of the above to the comment.

Also I've changed `NonNull<*mut c_void>` to `NonNull<c_void>` because by the docs `NonNull<T>` is '`*mut T` but non-null and covariant'

Pretty sure `NonZeroUsize` would also work but there are all of the opaque pointer intricacies that I am not fluent in so keeping it like that